Add support methods for AVIF changes to parser

baumanj opened the pull request:
See https://github.com/mozilla/mp4parse-rust/pull/193

baumanj committed:
Add FallibleVec::try_extend_from_slice

Also, check for overflowing the capacity on try_reserve and add tests.

baumanj committed:
Fix over-allocation in FallibleVec::try_reserve

Previously, the argument (named `new_cap`) was unconditionally added to the
existing capacity and the storage was reallocated. This seems to be
inconsistent with existing usage (and the Vec::reserve interface), which
call reserve with the number of additional elements which are about to be
added to the collection, presumably to avoid multiple allocations.

This changes the name of the argument to `additional` and changes the logic
to allocate space for that many additional insertions, considering the
current capacity, and not allocating unnecessarily.

Also, add some additional tests and remove the superfluous `_test` suffix
since test functions are already tagged with `#[test]` attributes.

baumanj committed:
Add try_read_to_end and tests

kinetiknz commented on lib.rs:
```diff
@@ -28,6 +30,27 @@ pub trait FallibleVec<T> {
+/// to read would have succeeded. In general, it is assumed that the callers
+/// have accurate knowledge of the number of bytes of interest and have created
+/// `src` accordingly.
+pub fn try_read_to_end<T>(src: &mut std::io::Take<T>, buf: &mut Vec<u8>) -> Result<usize, ()>
```
This looks like a fine approach, but I think it should live inside mp4parse-rust (in the same way allocate_read_buf does), since it's built on FallibleVec::try_reserve and this crate is trying to provide the minimum possible in terms of implementation - ideally we will be able to drop this crate and use Rust's fallible allocations directly.

kinetiknz reviewed (approved):
Nice find in try_reserve!  This doesn't change behaviour since we always passed an empty Vec from mp4parse-rust, but it's good to fix anyway.

Other changes look good to merge, but I'd prefer to implement try_read_to_end inside mp4parse-rust, since it's just building on try_reserve from this crate.

baumanj commented:
I'll implement your feedback and upload a new revision. Though I'm going to go through the rest of your feedback on https://github.com/mozilla/mp4parse-rust/pull/193 first in case I need to make any additional changes to this crate.

baumanj committed:
Move try_read_to_end to mp4parse crate per PR feedback

Also, add try_reserve_idempotent test to check against the specific regression
in try_reserve.

baumanj commented:
Changes on this side are done; ready for re-review

kinetiknz commented:
Thanks!
